bayes_kit/drghmc.py

```diff
@@ -0,0 +1,470 @@
+from typing import Iterator, Optional, Tuple
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+from .typing import DrawAndLogP, GradModel, Seed, VectorType
+
+
+class DrGhmcDiag:
+    """Generalized HMC sampler with Probabilistic Delayed Rejection diagonal metric.
+
+    We seek to sample draws -- denoted by (theta, rho) -- from a target distribution
+    specified in the model. As an HMC-variant, theta represents position (model
+    parameters) and rho represents momentum.
+
+    To produce a new draw, we generate a proposed draw with Hamiltonian dynamics. If
+    accepted, the proposed draw and its log density are returned; if rejected, a new
+    proposal is generated. This process is repeated until a maximum number of proposals
+    is reached or when probabilistic delayed rejection returns early.
+
+    This sampler requires non-increasing leapfrog stepsizes such that subsequent
+    proposals are generated with (equal or) more accurate Hamiltonian dynamics. This
+    allows for excellent sampling from multiscale distributions: large stepsizes are
+    used in wide, flat regions while smaller stepsizes are used in narrow,
+    high-curvature regions.
+
+    Implementation based on Modi C, Barnett A, Carpenter B. Delayed rejection
+    Hamiltonian Monte Carlo for sampling multiscale distributions. Bayesian Analysis.
+    2023. https://doi.org/10.1214/23-BA1360.
+    """
+
+    def __init__(
+        self,
+        model: GradModel,
+        proposals: int,
+        leapfrog_stepsizes: list[float],
```

Comment:
We haven't completely settled on this, but previous discussions we have had lent us toward adaptive versions being unique classes and maybe sharing code between them in utility functions etc, rather than subclassing/wrapping directly

Reply:
Fair enough.  These days, I believe unique classes for unique algorithms is the way to go.  But then I struggle with the design decisions about which of the methods in DrGhmcDiag should be written as a utility functions.  For instance, I'd argue `leapfrog()` should be pulled out of this class for sure, but what about `accept()`?  And both make good use of `self._cache`.

Reply:
Hard to say until the sibling-algorithm is being written IMO. I think you could pass `cache` as an argument rather than relying on `self`, if needed.

This also isn't exactly the same kind of code-reuse, but if the adaptive algorithm has a warmup which completes at some point, then `myadaptivesampler.stop_adaptation()` could return a normal, non-adaptive, class like this